src/python/pants/backend/python/util_rules/lockfile_diff.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+import logging
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Mapping
+
+from packaging.version import parse
+
+if TYPE_CHECKING:
+    # We seem to get a version of `packaging` that doesn't have `LegacyVersion` when running
+    # pytest..
+    from packaging.version import LegacyVersion, Version
+
+from pants.backend.python.util_rules.pex_requirements import (
+    LoadedLockfile,
+    LoadedLockfileRequest,
+    Lockfile,
+    LockfileContent,
+)
+from pants.base.exceptions import EngineError
+from pants.core.goals.generate_lockfiles import LockfileDiff, LockfilePackages, PackageName
+from pants.engine.fs import Digest, DigestContents
+from pants.engine.rules import Get, rule_helper
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, order=True)
+class PythonRequirementVersion:
+    _parsed: LegacyVersion | Version
+
+    @classmethod
+    def parse(cls, version: str) -> PythonRequirementVersion:
+        return cls(parse(version))
+
+    def __str__(self) -> str:
+        return str(self._parsed)
+
+    def __getattr__(self, key: str) -> Any:
+        return getattr(self._parsed, key)
+
+
+def _pex_lockfile_requirements(lockfile_data: Mapping[str, Any] | None) -> LockfilePackages:
+    if not lockfile_data:
+        return LockfilePackages({})
+
+    # Setup generators
+    locked_resolves = (
+        (
+            (PackageName(r["project_name"]), PythonRequirementVersion.parse(r["version"]))
+            for r in resolve["locked_requirements"]
```

Comment:
It will not fail miserably.. I say as I realize I've overlooked one detail. The idea was that it was protected being a generator, and as such will not be evaluated until iterated over within the `try .. except` when creating the `requirements` dict; _but_ the initial `lockfile_data["locked_resolves"]` are evaluated during generator creation, so that could fail.

Might just as well move the whole thing inside the try block to avoid this nuance altogether. 

I had opted for a debug message to not make a big fuzz of lockfiles we don't recognize as that could turn out noisy. We do support using custom lockfiles and if that's the case it would be a nuance to warn about it every time we try to.. oh, of course we wouldn't be here unless we generated it. Hmmm... 🤷🏽  Could turn it into a warning if that feels more appropriate?

Reply:
Well, a custom lockfile would fail to parse here, but you've asked for a diff, so presumably we should warn that what you've asked for is not possible? You can make the warning go away by not requesting a diff, which we can't provide anyway...

But if we inspect the metadata version we can give a more specific warning, i.e., "this lockfile is version X and we only know how to produce diffs for version 3". But that's nice-to-have, not critical.

Reply:
Yea, I'll defer that for now. I think you won't end up here with custom lockfiles (as we don't generate custom lockfiles in the first place)